monitoring/monitoring/monitoring.py

```diff
@@ -0,0 +1,219 @@
+import datetime
+import calendar
+import asyncio
+from aiohttp import web
+import aiohttp_session
+import logging
+from collections import defaultdict
+
+from hailtop import aiogoogle
+from hailtop.aiogoogle import BigQueryClient
+from hailtop.config import get_deploy_config
+from hailtop.hail_logging import AccessLogger
+from hailtop.tls import get_in_cluster_server_ssl_context
+from hailtop.utils import run_if_changed_idempotent, retry_long_running, time_msecs
+from gear import (Database, setup_aiohttp_session,
+                  web_authenticated_developers_only, transaction)
+from web_common import (setup_aiohttp_jinja2, setup_common_static_routes,
+                        render_template, set_message)
+
+log = logging.getLogger('monitoring')
+
+routes = web.RouteTableDef()
+
+deploy_config = get_deploy_config()
+
+
+def get_previous_month(dt):
+    if dt.month == 1:
+        return datetime.datetime(dt.year - 1, 12, 1)
+    return datetime.datetime(dt.year, dt.month - 1, 1)
+
+
+def get_last_day_month(dt):
+    _, last_day = calendar.monthrange(dt.year, dt.month)
+    return datetime.datetime(dt.year, dt.month, last_day)
+
+
+def format_data(records):
+    cost_by_service = defaultdict(lambda: 0)
+    compute_cost_breakdown = defaultdict(lambda: 0)
+    cost_by_sku_label = []
+
+    for record in records:
+        cost_by_sku_label.append(record)
+
+        cost_by_service[record['service_description']] += record['cost']
+
+        if record['service_description'] == 'Compute Engine':
+            compute_cost_breakdown[record['source']] += record['cost']
+
+    cost_by_service = sorted([{'service': k, 'cost': v} for k, v in cost_by_service.items()], key=lambda x: x['cost'], reverse=True)
+    compute_cost_breakdown = sorted([{'source': k, 'cost': v} for k, v in compute_cost_breakdown.items()], key=lambda x: x['cost'], reverse=True)
+    cost_by_sku_label.sort(key=lambda x: x['cost'], reverse=True)
+
+    return (cost_by_service, compute_cost_breakdown, cost_by_sku_label)
+
+
+async def _billing(request):
+    app = request.app
+    date_format = '%m/%Y'
+
+    now = datetime.datetime.now()
+    default_time_period = now.strftime(date_format)
+
+    time_period_query = request.query.get('time_period', default_time_period)
+
+    async def parse_error(msg):
+        session = await aiohttp_session.get_session(request)
+        set_message(session, msg, 'error')
+        return ([], [], [], time_period_query)
+
+    try:
+        time_period = datetime.datetime.strptime(time_period_query, date_format)
+        month = time_period.month
+        year = time_period.year
+        if time_period < datetime.datetime.strptime('07/2020', date_format):
+            return await parse_error(f"Invalid value for time_period '{time_period_query}'; time periods before 07/2020 are not supported.")
+    except ValueError:
+        return await parse_error(f"Invalid value for time_period '{time_period_query}'; must be in the format of MM/YYYY.")
```

Comment:
I think you intend to only catch ValueErrors coming from the first line of the try, right? I'd prefer we tighten the try-except to just wrap that line.